### PR TITLE
upgrade gsclient-go & clean up code and docs & add server datasource

### PR DESCRIPTION
**Pull request by nvthongswansea**
- Add missing server datasource and its docs.
- Handle all errors when setting values.
- Fix schemas (type mismatch, missing properties, etc.).
- Robust error reporting.
- Upgrade gsclient-go to v2.2.1
- Add table of feature matrix.

### CODE REVIEW

**Comment by itakouna** [comment]
@nvthongswansea please, it would be very helpful to make the PR short and target only one issue. This PR can be separated into 3 main issues: 

- Upgrade gsclient-go to v2.2.1 

- Add missing server datasource and its docs,  Fix schemas (type mismatch, missing properties, etc.).,  and Add table of feature matrix. 

- Handle all errors when setting values and Robust error reporting.

**Review by bkircher — APPROVED** [review]
Acceptance tests pass in this branch and changes look correct to me.

@itakouna 's comment is absolutely right, though. Please try to make smaller PRs. Try to split them up by topic and main issue so that we can follow the changes more easily. (I know its not easy sometimes, but is definitely much easier to follow the history and review, so in the end, it might even save time overall.)

**Comment by bkircher** [comment]
Question: Since a lot of changes in this PR are in vendor/ does it make sense that we include vendor dependencies in VCS? I am a Golang novice and thus might be wrong, but this feels bit odd.

A comment on the interwebs I could find says: don't: https://getcomposer.org/doc/faqs/should-i-commit-the-dependencies-in-my-vendor-directory.md ?

**Comment by bkircher** [comment]
Regarding vendor/, opinions seemingly vary. Found a [/r/golang/ thread](https://www.reddit.com/r/golang/comments/8s1cjd/do_i_gitignore_the_vendor_folder_dep/)  with some popular ones. (Not that Reddit is necessarily an authority). Anyway, just learned about this. Maybe we discuss this in another context, not this PR.

File: README.md [hunk 1]
```diff
@@ -76,6 +76,30 @@ In order to run a specific Acceptance test.
...
+| Firewall (CRUD) | :heavy_check_mark: |  
+| SSH key (CRUD) | :heavy_check_mark: |  
+| ISO Image (CRUD) | :heavy_check_mark: |  
+| Snapshot (CRUD) | :heavy_check_mark: |  
+| Snapshot rollback | :heavy_check_mark: |  
+| Snapshot to S3 | :x: |  
+| Snapshot schedule (CRUD) | :heavy_check_mark: |  
+| Template (CRUD) | :heavy_check_mark: |  
+| Multiple project support | :x: |
+
```
[the remark this answers is not in the document]
Comment (nvthongswansea) [reply README.md:102]: Sure

**Pull request closed without merging**